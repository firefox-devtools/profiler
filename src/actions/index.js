/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
