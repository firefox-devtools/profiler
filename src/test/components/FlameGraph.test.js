/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { fireEvent, within, screen } from '@testing-library/react';
import { Provider } from 'react-redux';

// This module is mocked.
import copy from 'copy-to-clipboard';

import { render, act } from 'firefox-profiler/test/fixtures/testing-library';
import { FlameGraph } from '../../components/flame-graph';
import { CallNodeContextMenu } from '../../components/shared/CallNodeContextMenu';
import {
  getInvertCallstack,
  getSourceViewFile,
} from '../../selectors/url-state';
import { StringTable } from '../../utils/string-table';
import { ensureExists } from '../../utils/flow';
import {
  getEmptyThread,
  getEmptyProfile,
} from '../../profile-logic/data-structures';
import {
  changeInvertCallstack,
  changeSelectedCallNode,
  commitRange,
  updatePreviewSelection,
  changeImplementationFilter,
} from '../../actions/profile-view';
import { changeSelectedTab } from '../../actions/app';
import { selectedThreadSelectors } from '../../selectors/per-thread';

import {
  autoMockCanvasContext,
  flushDrawLog,
} from '../fixtures/mocks/canvas-context';
import { storeWithProfile } from '../fixtures/stores';
import {
  addRootOverlayElement,
  removeRootOverlayElement,
  getMouseEvent,
  fireFullClick,
  fireFullContextMenu,
  findFillTextPositionFromDrawLog,
} from '../fixtures/utils';
import { getProfileFromTextSamples } from '../fixtures/profiles/processed-profile';
import { mockRaf } from '../fixtures/mocks/request-animation-frame';
import { autoMockElementSize } from '../fixtures/mocks/element-size';

import type { CssPixels } from 'firefox-profiler/types';

const GRAPH_WIDTH = 200;
const GRAPH_HEIGHT = 300;

describe('FlameGraph', function () {
  autoMockCanvasContext();
  autoMockElementSize({ width: GRAPH_WIDTH, height: GRAPH_HEIGHT });
  afterEach(removeRootOverlayElement);
  beforeEach(addRootOverlayElement);

  it('matches the snapshot', () => {
    const { container } = setupFlameGraph();
    const drawCalls = flushDrawLog();

    expect(container.firstChild).toMatchSnapshot();
    expect(drawCalls).toMatchSnapshot();
  });

  it('ignores invertCallstack and always displays non-inverted', () => {
    const { getState, dispatch } = setupFlameGraph();
    expect(getInvertCallstack(getState())).toBe(false);
    act(() => {
      dispatch(changeInvertCallstack(true));
    });
    expect(getInvertCallstack(getState())).toBe(false);
    act(() => {
      dispatch(changeInvertCallstack(false));
    });
    expect(getInvertCallstack(getState())).toBe(false);
  });

  it('shows a tooltip when hovering', () => {
    const { getTooltip, moveMouse, findFillTextPosition } = setupFlameGraph();
    expect(getTooltip()).toBe(null);
    moveMouse(findFillTextPosition('A'));
    expect(getTooltip()).toBeTruthy();
  });

  it('should not persist the selected frame tooltips', () => {
    const { getTooltip, moveMouse, findFillTextPosition, leftClick } =
      setupFlameGraph();
    // No tooltip displayed yet.
    expect(getTooltip()).toBe(null);

    leftClick(findFillTextPosition('A'));

    // The tooltip should be displayed.
    expect(getTooltip()).toBeTruthy();

    // Move the mouse outside of the frame.
    moveMouse({ x: 0, y: 0 });

    // Make sure that we don't have a persisted tooltip.
    expect(getTooltip()).toBeFalsy();
  });

  it('has a tooltip that matches the snapshot with categories', () => {
    const { getTooltip, moveMouse, findFillTextPosition } = setupFlameGraph();
    moveMouse(findFillTextPosition('A'));
    expect(getTooltip()).toMatchSnapshot();
  });

  it('shows a tooltip with the resource information with categories', () => {
    const { getTooltip, moveMouse, findFillTextPosition } = setupFlameGraph();
    moveMouse(findFillTextPosition('J'));
    const tooltip = ensureExists(getTooltip());

    // First, a targeted test.
    const { getByText } = within(tooltip);
    const resourceLabel = getByText('Resource:');
    const valueElement = ensureExists(resourceLabel.nextSibling);

    // See https://github.com/testing-library/jest-dom/issues/306
    // eslint-disable-next-line jest-dom/prefer-to-have-text-content
    expect(valueElement.textContent).toBe('libxul.so');
    // But also do a good old snapshot.
    expect(tooltip).toMatchSnapshot();
  });

  it('can be navigated with the keyboard', () => {
    const { getState, dispatch, getContentDiv, funcNames } = setupFlameGraph();
    const div = getContentDiv();

    function selectedNode() {
      const callNodeIndex =
        selectedThreadSelectors.getSelectedCallNodeIndex(getState());
      return callNodeIndex && funcNames[callNodeIndex];
    }

    // Start out with callnode B selected
    act(() => {
      dispatch(changeSelectedCallNode(0, [0, 1] /* B */));
    });
    expect(selectedNode()).toBe('B');

    // Move one callnode up
    fireEvent.keyDown(div, { key: 'ArrowUp' });
    expect(selectedNode()).toBe('C');

    // Move one callnode right
    fireEvent.keyDown(div, { key: 'ArrowRight' });
    expect(selectedNode()).toBe('H');

    // Go back to left again
    fireEvent.keyDown(div, { key: 'ArrowLeft' });
    expect(selectedNode()).toBe('C');

    // And down, back to our starting callnode again
    fireEvent.keyDown(div, { key: 'ArrowDown' });
    expect(selectedNode()).toBe('B');

    expect(getSourceViewFile(getState())).toBeNull();
    // Open source file for our starting callnode
    fireEvent.keyDown(div, { key: 'Enter' });
    expect(getSourceViewFile(getState())).toBe('path/for/B');
  });

  it('displays a context menu when rightclicking', () => {
    // Fake timers are indicated when dealing with the context menus.
    jest.useFakeTimers();

    const { rightClick, clickMenuItem, getContextMenu, findFillTextPosition } =
      setupFlameGraph();

    rightClick(findFillTextPosition('A'));
    expect(getContextMenu()).toHaveClass('react-contextmenu--visible');
    clickMenuItem('Copy function name');
    expect(copy).toHaveBeenLastCalledWith('A');

    // The right click gesture triggered 2 redraws (one mousemove, one
    // mousedown), but the tool findFillTextPosition ensures there's only one
    // result and will throw otherwise.
    // So let's flush the draw calls now. After running the timers
    // afterwards the flame graph will redraw again (as a result of closing the
    // menu and resetting the rightClickedCallNodeIndex).
    flushDrawLog();
    act(() => jest.runAllTimers());

    // Try another node to make sure the menu can handle other nodes than the first.
    rightClick(findFillTextPosition('B'));
    expect(getContextMenu()).toHaveClass('react-contextmenu--visible');
    clickMenuItem('Copy function name');
    expect(copy).toHaveBeenLastCalledWith('B');
  });

  it('has a tooltip that matches the snapshot with categories when a preview selection is applied', () => {
    const { getTooltip, moveMouse, findFillTextPosition, dispatch } =
      setupFlameGraph();
    flushDrawLog();
    act(() => {
      dispatch(
        updatePreviewSelection({
          hasSelection: true,
          isModifying: false,
          selectionStart: 1.3,
          selectionEnd: 5,
        })
      );
    });
    moveMouse(findFillTextPosition('A'));
    expect(getTooltip()).toMatchSnapshot();
  });

  describe('EmptyReasons', () => {
    it('matches the snapshot when a profile has no samples', () => {
      const profile = getEmptyProfile();
      const thread = getEmptyThread();
      thread.name = 'Empty Thread';
      profile.threads.push(thread);

      const store = storeWithProfile(profile);
      const container = render(
        <Provider store={store}>
          <>
            <FlameGraph />
          </>
        </Provider>
      ).container;

      expect(container.querySelector('.EmptyReasons')).toMatchSnapshot();
    });

    it('shows reasons when samples are not in the committed range', () => {
      const { dispatch } = setupFlameGraph();
      act(() => {
        dispatch(commitRange(5, 10));
      });
      expect(
        screen.getByText('Broaden the selected range to view samples.')
      ).toBeInTheDocument();
    });

    it('shows reasons when samples are not in the preview range', () => {
      const { dispatch } = setupFlameGraph();
      act(() => {
        dispatch(
          updatePreviewSelection({
            hasSelection: true,
            isModifying: false,
            selectionStart: 5,
            selectionEnd: 10,
          })
        );
      });

      expect(
        screen.getByText(
          'Try broadening the selected range, removing search terms, or call tree transforms to view samples.'
        )
      ).toBeInTheDocument();
    });

    it('shows reasons when samples have been completely filtered out', function () {
      const { dispatch } = setupFlameGraph();
      act(() => {
        dispatch(changeImplementationFilter('js'));
      });
      expect(
        screen.getByText(
          'Try broadening the selected range, removing search terms, or call tree transforms to view samples.'
        )
      ).toBeInTheDocument();
    });
  });
});

function setupFlameGraph() {
  const flushRafCalls = mockRaf();

  const {
    profile,
    funcNamesPerThread: [funcNames],
    funcNamesDictPerThread: [funcNamesDict],
  } = getProfileFromTextSamples(`
    A[cat:DOM]       A[cat:DOM]       A[cat:DOM]
    B[cat:DOM]       B[cat:DOM]       B[cat:DOM]
    C[cat:Graphics]  C[cat:Graphics]  H[cat:Network]
    D[cat:Graphics]  F[cat:Graphics]  I[cat:Network]
    E[cat:Graphics]  G[cat:Graphics]
                     J[lib:libxul.so]
  `);

  // Add some file and line number to the profile so that tooltips generate
  // an interesting snapshot.
  const { funcTable, stringArray } = profile.threads[0];
  const stringTable = StringTable.withBackingArray(stringArray);
  for (let funcIndex = 0; funcIndex < funcTable.length; funcIndex++) {
    funcTable.lineNumber[funcIndex] = funcIndex + 10;
    funcTable.columnNumber[funcIndex] = funcIndex + 100;
    funcTable.fileName[funcIndex] = stringTable.indexForString('path/to/file');
  }

  funcTable.fileName[funcNamesDict.B] =
    stringTable.indexForString('path/for/B');

  funcTable.fileName[funcNamesDict.J] = stringTable.indexForString(
    'hg:hg.mozilla.org/mozilla-central:widget/cocoa/nsAppShell.mm:997f00815e6bc28806b75448c8829f0259d2cb28'
  );

  const store = storeWithProfile(profile);
  store.dispatch(changeSelectedTab('flame-graph'));

  const renderResult = render(
    <Provider store={store}>
      <>
        <CallNodeContextMenu />
        <FlameGraph />
      </>
    </Provider>
  );
  const { container, getByText } = renderResult;

  flushRafCalls();

  function getPositioningOptions(x, y) {
    // These positioning options will be sent to all our mouse events. Note
    // that the values aren't really consistent, especially offsetY and
    // pageY shouldn't be the same, but in the context of our test this will
    // be good enough.
    // pageX/Y values control the position of the tooltip so it's not super
    // important.
    // offsetX/Y are more important as they're used to find which node is
    // actually clicked.
    // clientX/Y is used in the Viewport HOC when dragging and zooming.
    const positioningOptions = {
      offsetX: x,
      offsetY: y,
      clientX: x,
      clientY: y,
      pageX: x,
      pageY: y,
    };

    return positioningOptions;
  }

  const canvas = ensureExists(
    container.querySelector('canvas'),
    'The container should contain a canvas element.'
  );

  function fireMouseEvent(eventName, options) {
    fireEvent(canvas, getMouseEvent(eventName, options));
  }

  function leftClick({ x, y }: { x: CssPixels, y: CssPixels }) {
    const positioningOptions = getPositioningOptions(x, y);

    fireMouseEvent('mousemove', positioningOptions);
    fireFullClick(canvas, positioningOptions);
    flushRafCalls();
  }

  // You can use findFillTextPosition to derive the x, y positioning from the
  // draw log.
  function rightClick({ x, y }: { x: CssPixels, y: CssPixels }) {
    const positioningOptions = getPositioningOptions(x, y);

    fireMouseEvent('mousemove', positioningOptions);
    fireFullContextMenu(canvas, positioningOptions);
    flushRafCalls();
  }

  function moveMouse({ x, y }) {
    fireMouseEvent('mousemove', getPositioningOptions(x, y));
  }

  /**
   * The tooltip is in a portal, and created in the root overlay elements.
   */
  function getTooltip() {
    return document.querySelector('#root-overlay .tooltip');
  }

  /**
   * The content div is the one receiving keyboard events for navigation.
   */
  function getContentDiv() {
    return ensureExists(
      container.querySelector('.flameGraphContent'),
      `Couldn't find the content div with selector .flameGraphContent`
    );
  }

  // Context menu tools
  const getContextMenu = () =>
    ensureExists(
      container.querySelector('.react-contextmenu'),
      `Couldn't find the context menu.`
    );

  function clickMenuItem(strOrRegexp) {
    fireFullClick(getByText(strOrRegexp));
  }

  function findFillTextPosition(fillText: string) {
    return findFillTextPositionFromDrawLog(flushDrawLog(), fillText);
  }

  return {
    ...store,
    ...renderResult,
    funcNames,
    moveMouse,
    leftClick,
    rightClick,
    getTooltip,
    getContentDiv,
    getContextMenu,
    clickMenuItem,
    findFillTextPosition,
  };
}
