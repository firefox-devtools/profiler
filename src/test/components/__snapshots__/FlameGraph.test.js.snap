// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlameGraph EmptyReasons matches the snapshot when a profile has no samples 1`] = `
<div
  class="EmptyReasons"
>
  <h3>
    The 
    flame graph
     is empty for “
    Empty Thread
    ”
  </h3>
  <p>
    This thread has no samples.
  </p>
</div>
`;

exports[`FlameGraph has a tooltip that matches the snapshot with categories 1`] = `
<div
  class="tooltip"
  data-testid="tooltip"
  style="left: 14px; top: 306px;"
>
  <div
    class="tooltipCallNode"
  >
    <div
      class="tooltipOneLine tooltipHeader"
    >
      <div
        class="tooltipTiming"
      >
        3.0ms (100%)
      </div>
      <div
        class="tooltipTitle"
      >
        A
      </div>
      <div
        class="tooltipIcon"
      />
    </div>
    <div
      class="tooltipCallNodeDetails"
    >
      <div
        class="tooltipDetails tooltipCallNodeDetailsLeft"
      >
        <div
          class="tooltipLabel"
        >
          Stack Type:
        </div>
        <div>
          Native
        </div>
        <div
          class="tooltipLabel"
        >
          Category:
        </div>
        <div>
          <span
            class="colored-square category-color-blue"
          />
          DOM
        </div>
        <div
          class="tooltipLabel"
        >
          File:
        </div>
        <div
          class="tooltipDetailsUrl"
        >
          path/to/file:10:100
        </div>
      </div>
      <div
        class="tooltipCallNodeCategory"
      >
        <div />
        <div
          class="tooltipCallNodeHeader"
        />
        <div
          class="tooltipCallNodeHeader"
        >
          <span
            class="tooltipCallNodeHeaderSwatchRunning"
          />
          Running
        </div>
        <div
          class="tooltipCallNodeHeader"
        >
          <span
            class="tooltipCallNodeHeaderSwatchSelf"
          />
          Self
        </div>
        <div
          class="tooltipLabel tooltipCategoryLabel tooltipCategoryRowHeader"
        >
          Overall
        </div>
        <div
          class="tooltipCallNodeGraph tooltipCategoryRowHeader"
        >
          <div
            aria-label="Total samples overall"
            aria-valuemax="3"
            aria-valuemin="0"
            aria-valuenow="3"
            class="tooltipCallNodeGraphMeter tooltipCallNodeGraphMeterTotal"
            role="meter"
          >
            <div
              class="tooltipCallNodeGraphMeterBar"
              style="width: 100%;"
            />
          </div>
          <div
            aria-label="Self samples overall"
            aria-valuemax="3"
            aria-valuemin="0"
            aria-valuenow="0"
            class="tooltipCallNodeGraphMeter tooltipCallNodeGraphMeterSelf"
            role="meter"
          />
        </div>
        <div
          class="tooltipCallNodeTiming tooltipCategoryRow tooltipCategoryRowHeader"
        >
          3 samples
        </div>
        <div
          class="tooltipCallNodeTiming tooltipCategoryRow tooltipCategoryRowHeader"
        >
          —
        </div>
        <div
          class="tooltipCallNodeName tooltipLabel tooltipCategoryRowHeader"
        >
          Network
        </div>
        <div
          class="tooltipCallNodeGraph tooltipCategoryRowHeader"
        >
          <div
            aria-label="Total samples for Network"
            aria-valuemax="3"
            aria-valuemin="0"
            aria-valuenow="1"
            class="tooltipCallNodeGraphMeter tooltipCallNodeGraphMeterTotal"
            role="meter"
          >
            <div
              class="tooltipCallNodeGraphMeterBar"
              style="width: 33.33333333333333%;"
            />
          </div>
          <div
            aria-label="Self samples for Network"
            aria-valuemax="3"
            aria-valuemin="0"
            aria-valuenow="0"
            class="tooltipCallNodeGraphMeter tooltipCallNodeGraphMeterSelf"
            role="meter"
          />
        </div>
        <div
          class="tooltipCallNodeTiming tooltipCategoryRowHeader"
        >
          1 sample
        </div>
        <div
          class="tooltipCallNodeTiming tooltipCategoryRowHeader"
        >
          —
        </div>
        <div
          class="tooltipCallNodeName tooltipLabel tooltipCategoryRowHeader"
        >
          Graphics
        </div>
        <div
          class="tooltipCallNodeGraph tooltipCategoryRowHeader"
        >
          <div
            aria-label="Total samples for Graphics"
            aria-valuemax="3"
            aria-valuemin="0"
            aria-valuenow="2"
            class="tooltipCallNodeGraphMeter tooltipCallNodeGraphMeterTotal"
            role="meter"
          >
            <div
              class="tooltipCallNodeGraphMeterBar"
              style="width: 66.66666666666666%;"
            />
          </div>
          <div
            aria-label="Self samples for Graphics"
            aria-valuemax="3"
            aria-valuemin="0"
            aria-valuenow="0"
            class="tooltipCallNodeGraphMeter tooltipCallNodeGraphMeterSelf"
            role="meter"
          />
        </div>
        <div
          class="tooltipCallNodeTiming tooltipCategoryRowHeader"
        >
          2 samples
        </div>
        <div
          class="tooltipCallNodeTiming tooltipCategoryRowHeader"
        >
          —
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FlameGraph has a tooltip that matches the snapshot with categories when a preview selection is applied 1`] = `
<div
  class="tooltip"
  data-testid="tooltip"
  style="left: 14px; top: 306px;"
>
  <div
    class="tooltipCallNode"
  >
    <div
      class="tooltipOneLine tooltipHeader"
    >
      <div
        class="tooltipTiming"
      >
        1.0ms (100%)
      </div>
      <div
        class="tooltipTitle"
      >
        A
      </div>
      <div
        class="tooltipIcon"
      />
    </div>
    <div
      class="tooltipCallNodeDetails"
    >
      <div
        class="tooltipDetails tooltipCallNodeDetailsLeft"
      >
        <div
          class="tooltipLabel"
        >
          Stack Type:
        </div>
        <div>
          Native
        </div>
        <div
          class="tooltipLabel"
        >
          Category:
        </div>
        <div>
          <span
            class="colored-square category-color-blue"
          />
          DOM
        </div>
        <div
          class="tooltipLabel"
        >
          File:
        </div>
        <div
          class="tooltipDetailsUrl"
        >
          path/to/file:10:100
        </div>
      </div>
      <div
        class="tooltipCallNodeCategory"
      >
        <div />
        <div
          class="tooltipCallNodeHeader"
        />
        <div
          class="tooltipCallNodeHeader"
        >
          <span
            class="tooltipCallNodeHeaderSwatchRunning"
          />
          Running
        </div>
        <div
          class="tooltipCallNodeHeader"
        >
          <span
            class="tooltipCallNodeHeaderSwatchSelf"
          />
          Self
        </div>
        <div
          class="tooltipLabel tooltipCategoryLabel tooltipCategoryRowHeader"
        >
          Overall
        </div>
        <div
          class="tooltipCallNodeGraph tooltipCategoryRowHeader"
        >
          <div
            aria-label="Total samples overall"
            aria-valuemax="1"
            aria-valuemin="0"
            aria-valuenow="1"
            class="tooltipCallNodeGraphMeter tooltipCallNodeGraphMeterTotal"
            role="meter"
          >
            <div
              class="tooltipCallNodeGraphMeterBar"
              style="width: 100%;"
            />
          </div>
          <div
            aria-label="Self samples overall"
            aria-valuemax="1"
            aria-valuemin="0"
            aria-valuenow="0"
            class="tooltipCallNodeGraphMeter tooltipCallNodeGraphMeterSelf"
            role="meter"
          />
        </div>
        <div
          class="tooltipCallNodeTiming tooltipCategoryRow tooltipCategoryRowHeader"
        >
          1 sample
        </div>
        <div
          class="tooltipCallNodeTiming tooltipCategoryRow tooltipCategoryRowHeader"
        >
          —
        </div>
        <div
          class="tooltipCallNodeName tooltipLabel tooltipCategoryRowHeader"
        >
          Network
        </div>
        <div
          class="tooltipCallNodeGraph tooltipCategoryRowHeader"
        >
          <div
            aria-label="Total samples for Network"
            aria-valuemax="1"
            aria-valuemin="0"
            aria-valuenow="1"
            class="tooltipCallNodeGraphMeter tooltipCallNodeGraphMeterTotal"
            role="meter"
          >
            <div
              class="tooltipCallNodeGraphMeterBar"
              style="width: 100%;"
            />
          </div>
          <div
            aria-label="Self samples for Network"
            aria-valuemax="1"
            aria-valuemin="0"
            aria-valuenow="0"
            class="tooltipCallNodeGraphMeter tooltipCallNodeGraphMeterSelf"
            role="meter"
          />
        </div>
        <div
          class="tooltipCallNodeTiming tooltipCategoryRowHeader"
        >
          1 sample
        </div>
        <div
          class="tooltipCallNodeTiming tooltipCategoryRowHeader"
        >
          —
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FlameGraph matches the snapshot 1`] = `
<div
  aria-labelledby="flame-graph-tab-button"
  class="flameGraph"
  id="flame-graph-tab"
  role="tabpanel"
>
  <div
    class="stackSettings"
  >
    <ul
      class="panelSettingsList"
    >
      <li
        class="panelSettingsListItem"
      >
        <input
          checked=""
          class="photon-radio photon-radio-micro"
          id="implementation-radio-combined"
          title="Do not filter the stack frames"
          type="radio"
          value="combined"
        />
        <label
          class="photon-label photon-label-micro photon-label-horiz-padding"
          for="implementation-radio-combined"
          title="Do not filter the stack frames"
        >
          All frames
        </label>
        <input
          class="photon-radio photon-radio-micro"
          id="implementation-radio-js"
          title="Show only the stack frames related to JavaScript execution"
          type="radio"
          value="js"
        />
        <label
          class="photon-label photon-label-micro photon-label-horiz-padding"
          for="implementation-radio-js"
          title="Show only the stack frames related to JavaScript execution"
        >
          JavaScript
        </label>
        <input
          class="photon-radio photon-radio-micro"
          id="implementation-radio-cpp"
          title="Show only the stack frames for native code"
          type="radio"
          value="cpp"
        />
        <label
          class="photon-label photon-label-micro photon-label-horiz-padding"
          for="implementation-radio-cpp"
          title="Show only the stack frames for native code"
        >
          Native
        </label>
      </li>
    </ul>
    <div
      class="panelSearchField stackSettingsSearchField"
    >
      <label
        class="panelSearchFieldLabel"
      >
        Filter stacks: 
        <form
          class="idleSearchField panelSearchFieldInput"
        >
          <input
            class="idleSearchFieldInput photon-input"
            name="search"
            placeholder="Enter filter terms"
            required=""
            title="Only display stacks which contain a function whose name matches this substring"
            type="search"
            value=""
          />
          <input
            class="idleSearchFieldButton"
            tabindex="-1"
            type="reset"
          />
        </form>
      </label>
      <div
        class="panelSearchFieldIntroduction isHidden"
      >
        Did you know you can use the comma (,) to search using several terms?
      </div>
    </div>
  </div>
  <ol
    class="filterNavigatorBar calltreeTransformNavigator"
  >
    <li
      class="filterNavigatorBarItem filterNavigatorBarRootItem filterNavigatorBarSelectedItem filterNavigatorBarLeafItem"
    >
      <span
        class="filterNavigatorBarItemContent"
      >
        Complete “⁨Empty⁩”
      </span>
    </li>
  </ol>
  <div
    class="flameGraphContent"
  >
    <div
      class="react-contextmenu-wrapper treeViewContextMenu"
    >
      <div
        class="chartViewport"
        tabindex="0"
      >
        <div>
          <canvas
            class="chartCanvas flameGraphCanvas"
            height="300"
            style="width: 200px; height: 300px;"
            width="200"
          />
        </div>
        <div
          class="chartViewportScroll hidden"
        >
          Zoom Chart:
          <kbd
            class="chartViewportScrollKbd"
          >
            Ctrl + Scroll
          </kbd>
          or
          <kbd
            class="chartViewportScrollKbd"
          >
            Shift + Scroll
          </kbd>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FlameGraph matches the snapshot 2`] = `
Array [
  Array [
    "set font",
    "10px sans-serif",
  ],
  Array [
    "measureText",
    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.()< /:-_",
  ],
  Array [
    "measureText",
    "…",
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    0,
    200,
    300,
  ],
  Array [
    "set fillStyle",
    "#0060df60",
  ],
  Array [
    "fillRect",
    0,
    284,
    199.2,
    15,
  ],
  Array [
    "measureText",
    "A",
  ],
  Array [
    "set fillStyle",
    "#000",
  ],
  Array [
    "fillText",
    "A",
    3,
    295,
  ],
  Array [
    "set fillStyle",
    "#0060df60",
  ],
  Array [
    "fillRect",
    0,
    268,
    199.2,
    15,
  ],
  Array [
    "measureText",
    "B",
  ],
  Array [
    "set fillStyle",
    "#000",
  ],
  Array [
    "fillText",
    "B",
    3,
    279,
  ],
  Array [
    "set fillStyle",
    "#12bc0060",
  ],
  Array [
    "fillRect",
    0,
    252,
    133.2,
    15,
  ],
  Array [
    "measureText",
    "C",
  ],
  Array [
    "set fillStyle",
    "#000",
  ],
  Array [
    "fillText",
    "C",
    3,
    263,
  ],
  Array [
    "set fillStyle",
    "#45a1ff60",
  ],
  Array [
    "fillRect",
    134,
    252,
    65.19999999999999,
    15,
  ],
  Array [
    "measureText",
    "H",
  ],
  Array [
    "set fillStyle",
    "#000",
  ],
  Array [
    "fillText",
    "H",
    137,
    263,
  ],
  Array [
    "set fillStyle",
    "#12bc0060",
  ],
  Array [
    "fillRect",
    0,
    236,
    65.2,
    15,
  ],
  Array [
    "measureText",
    "D",
  ],
  Array [
    "set fillStyle",
    "#000",
  ],
  Array [
    "fillText",
    "D",
    3,
    247,
  ],
  Array [
    "set fillStyle",
    "#12bc0060",
  ],
  Array [
    "fillRect",
    66,
    236,
    67.19999999999999,
    15,
  ],
  Array [
    "measureText",
    "F",
  ],
  Array [
    "set fillStyle",
    "#000",
  ],
  Array [
    "fillText",
    "F",
    69,
    247,
  ],
  Array [
    "set fillStyle",
    "#45a1ff60",
  ],
  Array [
    "fillRect",
    134,
    236,
    65.19999999999999,
    15,
  ],
  Array [
    "measureText",
    "I",
  ],
  Array [
    "set fillStyle",
    "#000",
  ],
  Array [
    "fillText",
    "I",
    137,
    247,
  ],
  Array [
    "set fillStyle",
    "#12bc0060",
  ],
  Array [
    "fillRect",
    0,
    220,
    65.2,
    15,
  ],
  Array [
    "measureText",
    "E",
  ],
  Array [
    "set fillStyle",
    "#000",
  ],
  Array [
    "fillText",
    "E",
    3,
    231,
  ],
  Array [
    "set fillStyle",
    "#12bc0060",
  ],
  Array [
    "fillRect",
    66,
    220,
    67.19999999999999,
    15,
  ],
  Array [
    "measureText",
    "G",
  ],
  Array [
    "set fillStyle",
    "#000",
  ],
  Array [
    "fillText",
    "G",
    69,
    231,
  ],
  Array [
    "set fillStyle",
    "#12bc0060",
  ],
  Array [
    "fillRect",
    66,
    204,
    67.19999999999999,
    15,
  ],
  Array [
    "measureText",
    "J",
  ],
  Array [
    "set fillStyle",
    "#000",
  ],
  Array [
    "fillText",
    "J",
    69,
    215,
  ],
]
`;

exports[`FlameGraph shows a tooltip with the resource information with categories 1`] = `
<div
  class="tooltip"
  data-testid="tooltip"
  style="left: 80px; top: 226px;"
>
  <div
    class="tooltipCallNode"
  >
    <div
      class="tooltipOneLine tooltipHeader"
    >
      <div
        class="tooltipTiming"
      >
        1.0ms (33%)
      </div>
      <div
        class="tooltipTitle"
      >
        J
      </div>
      <div
        class="tooltipIcon"
      />
    </div>
    <div
      class="tooltipCallNodeDetails"
    >
      <div
        class="tooltipDetails tooltipCallNodeDetailsLeft"
      >
        <div
          class="tooltipLabel"
        >
          Stack Type:
        </div>
        <div>
          Native
        </div>
        <div
          class="tooltipLabel"
        >
          Category:
        </div>
        <div>
          <span
            class="colored-square category-color-green"
          />
          Graphics
        </div>
        <div
          class="tooltipLabel"
        >
          File:
        </div>
        <div
          class="tooltipDetailsUrl"
        >
          widget/cocoa/nsAppShell.mm:17:107
        </div>
        <div
          class="tooltipLabel"
        >
          Resource:
        </div>
        libxul.so
      </div>
      <div
        class="tooltipCallNodeCategory"
      >
        <div />
        <div
          class="tooltipCallNodeHeader"
        />
        <div
          class="tooltipCallNodeHeader"
        >
          <span
            class="tooltipCallNodeHeaderSwatchRunning"
          />
          Running
        </div>
        <div
          class="tooltipCallNodeHeader"
        >
          <span
            class="tooltipCallNodeHeaderSwatchSelf"
          />
          Self
        </div>
        <div
          class="tooltipLabel tooltipCategoryLabel tooltipCategoryRowHeader"
        >
          Overall
        </div>
        <div
          class="tooltipCallNodeGraph tooltipCategoryRowHeader"
        >
          <div
            aria-label="Total samples overall"
            aria-valuemax="1"
            aria-valuemin="0"
            aria-valuenow="1"
            class="tooltipCallNodeGraphMeter tooltipCallNodeGraphMeterTotal"
            role="meter"
          >
            <div
              class="tooltipCallNodeGraphMeterBar"
              style="width: 100%;"
            />
          </div>
          <div
            aria-label="Self samples overall"
            aria-valuemax="1"
            aria-valuemin="0"
            aria-valuenow="1"
            class="tooltipCallNodeGraphMeter tooltipCallNodeGraphMeterSelf"
            role="meter"
          >
            <div
              class="tooltipCallNodeGraphMeterBar"
              style="width: 100%;"
            />
          </div>
        </div>
        <div
          class="tooltipCallNodeTiming tooltipCategoryRow tooltipCategoryRowHeader"
        >
          1 sample
        </div>
        <div
          class="tooltipCallNodeTiming tooltipCategoryRow tooltipCategoryRowHeader"
        >
          1 sample
        </div>
        <div
          class="tooltipCallNodeName tooltipLabel tooltipCategoryRowHeader"
        >
          Graphics
        </div>
        <div
          class="tooltipCallNodeGraph tooltipCategoryRowHeader"
        >
          <div
            aria-label="Total samples for Graphics"
            aria-valuemax="1"
            aria-valuemin="0"
            aria-valuenow="1"
            class="tooltipCallNodeGraphMeter tooltipCallNodeGraphMeterTotal"
            role="meter"
          >
            <div
              class="tooltipCallNodeGraphMeterBar"
              style="width: 100%;"
            />
          </div>
          <div
            aria-label="Self samples for Graphics"
            aria-valuemax="1"
            aria-valuemin="0"
            aria-valuenow="1"
            class="tooltipCallNodeGraphMeter tooltipCallNodeGraphMeterSelf"
            role="meter"
          >
            <div
              class="tooltipCallNodeGraphMeterBar"
              style="width: 100%;"
            />
          </div>
        </div>
        <div
          class="tooltipCallNodeTiming tooltipCategoryRowHeader"
        >
          1 sample
        </div>
        <div
          class="tooltipCallNodeTiming tooltipCategoryRowHeader"
        >
          1 sample
        </div>
      </div>
    </div>
  </div>
</div>
`;
